dagshub/data_engine/model/query_result.py

```diff
@@ -268,31 +267,49 @@ def get_blob_fields(
-            cache_location = self.datasource.default_dataset_location / ".metadata_blobs"
-
-            if cache_on_disk:
-                cache_location.mkdir(parents=True, exist_ok=True)
-
-            blob_urls = list(map(lambda dp: dp._extract_blob_url_and_path(fld), self.entries))
-            auth = self.datasource.source.repoApi.auth
-            func_args = zip(
-                map(lambda bu: bu[0], blob_urls),
-                map(lambda bu: bu[1], blob_urls),
-                repeat(auth),
-                repeat(cache_on_disk),
-                repeat(load_into_memory),
+            fields = tuple(
+                [field.name for field in self.datasource.fields if field.valueType == MetadataFieldType.BLOB]
             )
-            with multiprocessing.pool.ThreadPool(num_proc) as pool:
-                res = pool.starmap(_get_blob, func_args)
 
-            for dp, binary_val_or_path in zip(self.entries, res):
-                if binary_val_or_path is None:
+        if len(fields) == 0:
+            logger.warning("No blob fields loaded")
+            return self
+
+        cache_location = self.datasource.default_dataset_location / ".metadata_blobs"
+
+        if cache_on_disk:
+            cache_location.mkdir(parents=True, exist_ok=True)
+
+        # Create a list of things to download: (datapoint, field, url, path_location)
+        to_download: List[Tuple[Datapoint, str, str, Path]] = []
+        for dp in self.entries:
+            for fld in fields:
+                field_value = dp.metadata.get(fld)
+                # If field_value is a blob or a path, then ignore, means it's already been downloaded
+                if not isinstance(field_value, str):
                     continue
-                dp.metadata[fld] = binary_val_or_path
+                download_task = (dp, fld, dp.blob_url(field_value), dp.blob_cache_location / field_value)
+                to_download.append(download_task)
+
+        progress = get_rich_progress(rich.progress.MofNCompleteColumn())
+        task = progress.add_task("Downloading blobs...", total=len(to_download))
+
+        auth = self.datasource.source.repoApi.auth
+
+        def _get_blob_fn(dp: Datapoint, field: str, url: str, blob_path: Path):
+            blob_or_path = _get_blob(url, blob_path, auth, cache_on_disk, load_into_memory)
+            if isinstance(blob_or_path, str):
+                logger.warning(f"Error while downloading blob for field {field} in datapoint {dp.path}:{blob_or_path}")
+            dp.metadata[field] = blob_or_path
+
+        with progress:
+            with ThreadPoolExecutor(max_workers=num_proc) as tp:
+                futures = [tp.submit(_get_blob_fn, *args) for args in to_download]
+                for f in as_completed(futures):
+                    exc = f.exception()
+                    if exc is not None:
+                        logger.warning(f"Got exception {type(exc)} while downloading blob: {exc}")
```

Comment:
Maybe the future can yield the actual task so that you can print info on the specific datapoint and blob?

Reply:
Most of this will be handled by the exception handler in `_get_blob_fn`, where the datapoint will be output as well as the field for which it happened.
This exception handler is more of a doomsday handler for really weird errors.

Besides, if `f.exception()` is not None, then that means that `f` doesn't have a result set, so there's no way for me to know which dp it was working on (unless I store the arguments along with the futures)